File: 05g-kafka-cloudevents-avro-with-springboot/src/python-cloudevents-kafka-producer-ce-native/cloud_event_producer.py
import uuid
from datetime import datetime, timezone

from cloudevents.v1.http import CloudEvent
from cloudevents.v1.kafka import to_binary
from confluent_kafka import Producer
from confluent_kafka.schema_registry import SchemaRegistryClient
from confluent_kafka.schema_registry.avro import AvroSerializer
from confluent_kafka.serialization import SerializationContext, MessageField


class CloudEventProducer:

    def __init__(self, bootstrap_servers: str, schema_registry_url: str, source: str):
        self._source = source
        self._schema_registry_url = schema_registry_url
        self._schema_registry_client = SchemaRegistryClient({"url": schema_registry_url})
        self._producer = Producer({"bootstrap.servers": bootstrap_servers})
        self._serializer_cache: dict[str, AvroSerializer] = {}

    def publish(
        self,
        topic: str,
        key: str,
        payload: dict,
        schema_str: str,
        event_type: str,
        subject_name: str | None = None,
    ) -> None:
        value_bytes = self._get_serializer(schema_str)(
            payload, SerializationContext(topic, MessageField.VALUE)
        )

        attributes = {
            "type": event_type,
            "source": self._source,
            "id": str(uuid.uuid4()),
            "time": datetime.now(timezone.utc).isoformat(),
            "datacontenttype": "avro/binary",
        }
        if subject_name:
            attributes["dataschema"] = (
                f"{self._schema_registry_url}/subjects/{subject_name}/versions/latest"
            )

        event = CloudEvent(attributes=attributes, data=value_bytes)
        message = to_binary(event, data_marshaller=lambda x: x)

        self._producer.produce(
            topic=topic,
            key=key,
            value=message.value,
            headers=message.headers,
            on_delivery=_delivery_callback,
        )
        self._producer.poll(0)

    def flush(self) -> None:
        self._producer.flush()

    def _get_serializer(self, schema_str: str) -> AvroSerializer:
        if schema_str not in self._serializer_cache:
            self._serializer_cache[schema_str] = AvroSerializer(
                self._schema_registry_client, schema_str
            )
        return self._serializer_cache[schema_str]


def _delivery_callback(err, msg) -> None:
    if err:
        print(f"ERROR: delivery failed [key={msg.key()}]: {err}")
    else:
        print(
            f"Published CE [key={msg.key()}, topic={msg.topic()}, "
            f"partition={msg.partition()}, offset={msg.offset()}]"
        )

File: 05g-kafka-cloudevents-avro-with-springboot/src/spring-boot-cloudevents-kafka-producer-ce-native/src/main/java/com/trivadis/kafkaws/events/CloudEventProducer.java
package com.trivadis.kafkaws.events;

import io.cloudevents.CloudEvent;
import io.cloudevents.core.builder.CloudEventBuilder;
import io.confluent.kafka.schemaregistry.client.CachedSchemaRegistryClient;
import io.confluent.kafka.schemaregistry.client.SchemaRegistryClient;
import io.confluent.kafka.serializers.KafkaAvroSerializer;
import io.confluent.kafka.serializers.KafkaAvroSerializerConfig;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.apache.avro.io.DatumWriter;
import org.apache.avro.io.Encoder;
import org.apache.avro.io.EncoderFactory;
import org.apache.avro.specific.SpecificDatumWriter;
import org.apache.avro.specific.SpecificRecord;
import org.apache.kafka.clients.producer.ProducerRecord;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.kafka.core.KafkaTemplate;
import org.springframework.kafka.support.SendResult;
import org.springframework.stereotype.Component;

import java.io.ByteArrayOutputStream;
import java.io.IOException;
import java.io.UncheckedIOException;
import java.net.URI;
import java.time.OffsetDateTime;
import java.util.HashMap;
import java.util.Map;
import java.util.UUID;
import java.util.concurrent.CompletableFuture;

@Slf4j
@Component
@RequiredArgsConstructor
public class CloudEventProducer {

    @Value("${app.kafka.schema-registry-url}")
    private String schemaRegistryUrl;

    @Value("${app.kafka.topics.orders}")
    private String kafkaTopic;

    private final KafkaTemplate<String, CloudEvent> kafkaTemplate;

    @Value("${cloudevents.source}")
    private String defaultSource;

    public <T extends SpecificRecord> CompletableFuture<SendResult<String, CloudEvent>> publish(
            String topic,
            String key,
            T payload,
            String eventType) {

        String subjectName = kafkaTopic + "-value";

        CloudEvent event = CloudEventBuilder.v1()
            .withId(UUID.randomUUID().toString())
            .withSource(URI.create(defaultSource))
            .withType(eventType)
            .withTime(OffsetDateTime.now())
            .withDataContentType("avro/binary")
            .withDataSchema(URI.create(schemaRegistryUrl + "/subjects/" + subjectName + "/versions/latest"))
            .withSubject("subject")
            .withExtension("partitionkey", key)
            .withData(toAvroBytesSR(payload))
            .build();

        String messageKey = (String) event.getExtension("partitionkey");
        ProducerRecord<String, CloudEvent> record =
            new ProducerRecord<>(topic, key, event);

        return kafkaTemplate.send(record)
            .whenComplete((result, ex) -> {
                if (ex != null) {
                    log.error("Failed to publish CE [type={}, key={}]: {}",
                        eventType, key, ex.getMessage(), ex);
                } else {
                    log.info("Published CE [type={}, key={}, partition={}, offset={}]",
                        eventType, key,
                        result.getRecordMetadata().partition(),
                        result.getRecordMetadata().offset());
                }
            });
    }

    /**
     * Version without schema registry support 
     */
    private <T extends SpecificRecord> byte[] toAvroBytes(T record) {
        try {
            ByteArrayOutputStream out = new ByteArrayOutputStream();
            DatumWriter<T> writer = new SpecificDatumWriter<>(record.getSchema());
            Encoder encoder = EncoderFactory.get().binaryEncoder(out, null);
            writer.write(record, encoder);
            encoder.flush();
            return out.toByteArray();
        } catch (IOException e) {
            throw new UncheckedIOException(e);
        }
    }

    private <T extends SpecificRecord> byte[] toAvroBytesSR(T record) {
        SchemaRegistryClient schemaRegistryClient = new CachedSchemaRegistryClient(schemaRegistryUrl, 10);
        Map<String, Object> props = new HashMap<>();
        // send correct schemas to the registry, without "avro.java.string"
        props.put(KafkaAvroSerializerConfig.AVRO_REMOVE_JAVA_PROPS_CONFIG, true);
        props.put("schema.registry.url", schemaRegistryUrl);
        KafkaAvroSerializer ser = new KafkaAvroSerializer(schemaRegistryClient, props);
        return ser.serialize(kafkaTopic, record);
    }    
}
